### PR TITLE
feat(components): add merge functionality for components

### PR DESCRIPTION
**Pull request by imaykay**
* added an abstract js module that is able to provide wizard functionality
* added a slightly more specific js module that provides some functionality needed to display a simple merge gui
* added a merge wizard for components that can be used by admins
  * used the above js modules as a framework, so wizard is completely js/ajax/json based
* added a new merge method in the components service because the standard update method does not update all fields
* please review and work on the FIXME comments :-)
* merging of releases is still open but might be implemented almost the same as merging of components

closes sw360/sw360portal#278

### CODE REVIEW

**Comment by imaykay** [comment]
Beside having a look on the new FIXMEs in DatabaseComponentHandler#mergeComponents, there seems to be at least one bug: the componentId is not updated inside migrated releases and I couldn't figure out in the remaining time, why this happens.

If there are any questions, just ask :-)

**Comment by mcjaeger** [comment]
@bodetc @alexbrdn I think on the telco on wednesday we did not check the fixme comments. What would be the approach for these?

**Comment by bodetc** [comment]
I am fixing them right now, then I guess Alex can review my changes :)

**Comment by mcjaeger** [comment]
@bodetc actually, .... did you review the code?

in case you did this, could you please add your review using the github review functionality? Otherwise we dod not have any clue whether you have reviewed the code or not.

**Comment by mcjaeger** [comment]
@bodetc OK, so you are finished so far or are you planning to continue working on this?

Could you please help me and point out open points in the conversation.

**Comment by bodetc** [comment]
I am done so far, as a limitation, the feature currently only merges components and moves all underlying releases for one to the other. Implementation of a release merge feature would be much more time intensive, as we need to look for all relations (vulnerabilities, projects, ...) However, the merge wizard implemented here can be reused to create a merge release functionality.

There is a few FIXME wich I could not answer:
// FIXME: also delete attachment contents of deleted attachments in target?
-> I think this would be done automagically, but I a not sure
// FIXME: where else are componentids referenced?
-> I didn't think of other direct usages (i.e. not via release)
// FIXME: what is document state for?
-> ?
// FIXME: maybe validate the component
-> Do we validate in other cases?

**Comment by mcjaeger** [comment]
thanks for pointing this out:
* valiate is somewhere when create is done. may the merged component can be seen as creation step and thus validate could be good. But since both parts of it went through validate, maybe it would be redundant.
* Searching the code yields the documentState indicated open moderation requests (https://github.com/sw360/sw360portal/blob/f7141d0309df51b1f0bd1d752701b4e844dd2c96/libraries/lib-datahandler/src/main/thrift/sw360.thrift) I thinkn that should be checked that there are no open moderation requests for the merged components. Otherwise we would have orphaned moderation requests.
* Moderation requests and releases are maybe besides the attachments another corner to look for references -> buy maybe we can close this for now, except for moderation requests.
* @bodetc  Automagic deletion of attachment contents: there is the database sanitation for this you mean? Or what did you have in mind?

File: frontend/sw360-portlet/src/main/webapp/js/modules/mergeWizard.js [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+            /* origVal should be kept but has been empty*/
+            return undefined;
+        } else {
+            /* origVal should be kept */
+            return origVal;
+        }
+    }
+
+    return mergeWizard;
+});
```
Comment (alexbrdn) [inline frontend/sw360-portlet/src/main/webapp/js/modules/mergeWizard.js]: missing newline at EOF

File: frontend/sw360-portlet/src/main/webapp/js/modules/sw360Wizard.js [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+     *         </div>
+     *     </div>
+     * </div>
+     *
+     * 2. You can then grab the wizard framework via require and configure it inside like that:
+     * require([ 'modules/sw360Wizard', 'jquery' ], function(wizard, $) {
+     *     wizard({
+     *         wizardRoot: $('#myWizard'),
+     *         postUrl: your-post-url-goes-here,
+     *         postParamsPrefix: your-post-params-prefix-goes-here,
```
Comment (alexbrdn) [inline frontend/sw360-portlet/src/main/webapp/js/modules/sw360Wizard.js]: I think calling it something like 'httpPostParamsPrefix' would be a more direct hint at what this parameter does. 
Although it is explained well enough in the long text below, so renaming is really optional.

File: frontend/sw360-portlet/src/main/webapp/js/modules/sw360Wizard.js [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            });
+
+            return reworkedPostData;
+        }
+
+        function fillContentOfStep(activeElement, activeIndex, nextElement, nextIndex) {
+            var proceed = true,
+                proceedTmp;
+
+            if (nextElement[0] === firstStep[0]) {
```
Comment (alexbrdn) [inline frontend/sw360-portlet/src/main/webapp/js/modules/sw360Wizard.js]: I'd like to see this unboxing of single-element arrays happen right where the search for them is performed with `$(...)`. It looks a little odd that variables named in singular (e.g. "nextElemen**t**") happen to be arrays.

File: frontend/sw360-portlet/src/main/webapp/js/modules/sw360Wizard.js [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        });
+
+        return {
+                // public API
+        };
+    }
+
+    return sw360Wizard;
+
+});
```
Comment (alexbrdn) [inline frontend/sw360-portlet/src/main/webapp/js/modules/sw360Wizard.js]: missing newline at EOF

**Comment by mcjaeger** [comment]
(from discussion)

* Moderation requests and releases are maybe besides the attachments another corner to look for references -> buy maybe we can close this for now, except for moderation requests.
-> resolution is to block merging of components, if open moderation requests  for either of the two components exist


@bodetc Automagic deletion of attachment contents: there is the database sanitation for this you mean? Or what did you have in mind?
-> resolution: testing required: maybe updateComponent (see comment on code review) does the job already.

**Comment by alexbrdn** [comment]
After long wait, I think this PR is finally ready to be merged.

**Comment by maxhbr** [comment]
Aggregated data (e.g. Licenses on the level of components) should be recalculated instead of merged by hand:

With this feature one is able to generate a component which does not list all child-licenses.

**Comment by mcjaeger** [comment]
Ok, good point, but I would rather ses this as an issue rather than going again into some extra round of adding some functionality, testing, code review, etc pp. because basically the merge feature shall be used by experts and not by Jon Doe users who are maybe prone to mess up things.  With the entire merge feature, the user is able to create a new data set which does not contain all information from the sources.

**Comment by mcjaeger** [comment]
@bodetc btw, how is about your requested changes?

**Comment by maxhbr** [comment]
Maybe unrelated and also present on master:

when opening component details: `jquery-confirm` is loaded without loading `jQuery` beforehand:

**Comment by maxhbr** [comment]
Minor GUI issue:

in a field with multiple merge options the background is not consistently set to orange. By toggling the second option one can always reset the background to white. 
### screenshots:

after toggling the first line:

**Comment by maxhbr** [comment]
# This does not update the reverse link for a release to its component
If a release is merged into another component, the field `componentId` of that release is unchanged.

## To verify:
This breaks the project view if a release which is used in a project is merged into another component.
### setup:
- Create component `CompA`
- Create component `CompA'`
- Create Release `RelA` of `CompA'`
- Use `RelA` as linked release in some project `ProjA`
### What to do
- Merge `CompA` and `CompA'` and add `RelA` to `CompA'`
### Then, when viewing the detail page of `ProjA`:

with the following log output:
```
[...]                                                                                   
sw360                  |  2017-12-05 17:01:08 INFO  DatabaseConnector:136 - Document not found for ID: b0e08556fcb3ea831201e20387002226                                                     
sw360                  |  2017-12-05 17:01:08 INFO  DatabaseConnector:136 - Document not found for ID: b0e08556fcb3ea831201e20387002226                                                     
sw360                  |  2017-12-05 17:01:08 ERROR Sw360ThriftServlet:44 - uncaught                                                                                                        
sw360                  |  java.lang.NullPointerException                                                                                                                                    
sw360                  |        at org.eclipse.sw360.datahandler.db.ProjectDatabaseHandler.getReleaseClearingStatuses(ProjectDatabaseHandler.java:504)                                      
sw360                  |        at org.eclipse.sw360.projects.ProjectHandler.getReleaseClearingStatuses(ProjectHandler.java:229)                                                            
sw360                  |        at org.eclipse.sw360.datahandler.thrift.projects.ProjectService$Processor$getReleaseClearingStatuses.getResult(ProjectService.java:2273)                    
sw360                  |        at org.eclipse.sw360.datahandler.thrift.projects.ProjectService$Processor$getReleaseClearingStatuses.getResult(ProjectService.java:2258)                    
sw360                  |        at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)                                                                                       
sw360                  |        at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)                                                                                         
sw360                  |        at org.apache.thrift.server.TServlet.doPost(TServlet.java:83)                                                                                               
sw360                  |        at org.eclipse.sw360.projects.Sw360ThriftServlet.doPost(Sw360ThriftServlet.java:42)                                                                         
sw360                  |        at javax.servlet.http.HttpServlet.service(HttpServlet.java:650)                                                                                             
sw360                  | 2017-12-05 17:01:08,237 ERROR ProjectPortlet:733 - Error fetching project from backend!                                                                            
sw360                  |  org.apache.thrift.transport.TTransportException: HTTP Response code: 500                                                                                          
sw360                  |        at org.apache.thrift.transport.THttpClient.flush(THttpClient.java:349)                                                                                      
sw360                  |        at org.apache.thrift.TServiceClient.sendBase(TServiceClient.java:73)                                                                                        
sw360                  |        at org.apache.thrift.TServiceClient.sendBase(TServiceClient.java:62)                                                                                        
sw360                  |        at org.eclipse.sw360.datahandler.thrift.projects.ProjectService$Client.send_getReleaseClearingStatuses(ProjectService.java:889)                             
sw360                  |        at org.eclipse.sw360.datahandler.thrift.projects.ProjectService$Client.getReleaseClearingStatuses(ProjectService.java:880)                                  
sw360                  |        at org.eclipse.sw360.portal.portlets.FossologyAwarePortlet.putReleasesAndProjectIntoRequest(FossologyAwarePortlet.java:207)                                 
sw360                  |        at org.eclipse.sw360.portal.portlets.projects.ProjectPortlet.prepareDetailView(ProjectPortlet.java:724)                                                     
sw360                  |        at org.eclipse.sw360.portal.portlets.projects.ProjectPortlet.doView(ProjectPortlet.java:604)                                                                
sw360                  |        at com.liferay.portal.kernel.portlet.LiferayPortlet.doDispatch(LiferayPortlet.java:235)                                                                     
sw360                  | 2017-12-05 17:01:08,246 ERROR detail_jsp:440 - Error in useBeans: java.lang.InstantiationException: bean numberOfUncheckedVulnerabilities not found within scope
```

**Comment by alexbrdn** [comment]
@maxhbr, thanks for the heads up. I believe I fixed the problem with `componentId`. See here https://github.com/sw360/sw360portal/pull/602/files#diff-13d2e9cdc0b86428d73e716a93671652R457

**Comment by alexbrdn** [comment]
removed aggregate fields from the merge wizard.

@maxhbr, please review

**Comment by maxhbr** [comment]
**Minor issue**: one is able to merge a component into itself. That is confusing and should be disabled (later).

**Pull request merged**